src/test/java/com/ita/if103java/ims/service/WarehouseServiceImplTest.java

```diff
@@ -0,0 +1,454 @@
+//         //form a return DTO
+        when(warehouseDtoMapper.toDto(warehouseCreate)).thenReturn(topWarehouseDto);
+        topWarehouseDto.setAddressDto(addressDto);
+
+    }
+
+    @Test
+    public void addWarehouse_testMaxWarehouseNumberReached() {
+        int quantity = 3;
+        warehouseDto.setParentID(null);
+        Long accountId = topWarehouseDto.getAccountID();
+        assertNull(topWarehouseDto.getParentID());
+        when(warehouseDao.findQuantityOfWarehousesByAccountId(1L)).thenReturn(quantity);
+
+        MaxWarehousesLimitReachedException exception = assertThrows(MaxWarehousesLimitReachedException.class, () -> {
+            warehouseService.add(warehouseDto, userDetails);
+        });
+        assertEquals("The maximum number of warehouses has been reached for this" + "{accountId = " + accountId + "}", exception.getMessage());
+    }
+
+    @Test
+    public void addWarehouse_testMaxDepthReached() {
+        Warehouse warehouse = new Warehouse(1L, "Bottom", "auto parts", 20, true, 5L, 1L, 4L, true);
+        int level = 4;
+        warehouseDto.setParentID(5L);
+        accountId = userDetails.getUser().getAccountId();
+        int maxDepth = userDetails.getAccountType().getMaxWarehouseDepth();
+        assertNotNull(warehouseDto.getParentID());
+        when(warehouseDao.findLevelByParentID(warehouseDto.getParentID())).thenReturn(level);
+        assertFalse(level < maxDepth);
+//        MaxWarehousesLimitReachedException exception = assertThrows(MaxWarehousesLimitReachedException.class, () -> {
+//            warehouseService.add(warehouseDto, userDetails);
+//        });
+//        assertEquals("The maximum number of warehouses has been reached for this" + "{accountId = " + accountId + "}", exception.getMessage());
+
+    }
+
+    @Test
+    void findByIdTest_isWarehouseTopLevel() {
+        Warehouse targetWarehouse = new Warehouse();
+        WarehouseDto warehouseDto = new WarehouseDto();
+        AddressDto addressDto = new AddressDto();
+        Address address = new Address();
+        when(warehouseDao.findById(1L, 1L)).thenReturn(targetWarehouse);
+        when(warehouseDtoMapper.toDto(targetWarehouse)).thenReturn(warehouseDto);
+        assertTrue(targetWarehouse.isTopLevel());
+        when(addressDao.findByWarehouseId(1L)).thenReturn(address);
+        when(addressDtoMapper.toDto(address)).thenReturn(addressDto);
+
+        assertEquals(warehouseService.findById(1L, userDetails), warehouseDto);
+    }
+
+    @Test
+    void findByIdTest_isNotWarehouseTopLevel() {
+        Warehouse bottom = new Warehouse(12L, "Bottom", "auto parts", 20, true, 5L, 1L, 4L, true);
+
+        when(warehouseDao.findById(1L, 1L)).thenReturn(bottom);
+        when(warehouseDtoMapper.toDto(bottom)).thenReturn(warehouseDto);
+        assertFalse(bottom.isTopLevel());
+
+        //populate path
+        when(warehouseDao.findByTopWarehouseID(2L, 1L)).thenReturn(Collections.emptyList());
+    }
```

Comment:
this test test nothing, there is no invocation of the warehouseService
Fix or remove